### PR TITLE
Add per-vertex attribute support to vulkan, spir-v and wgsl.

### PR DESCRIPTION
**Pull request by atlv24**
**Connections**
#8591

**Description**
HLSL and MSL support are possible, but hard. I want to land Vulkan support and worry about the rest later.

**Testing**
wgpu-gpu test and naga test.

**Squash or Rebase?**

squish

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry.

### CODE REVIEW

**Comment by inner-daemons** [comment]
I may work on the HLSL and MSL support myself at a later date. I'll be reviewing this PR, since I've examined the spec and talked about this briefly with @atlv24.

File: naga/src/valid/mod.rs [hunk 1]
```diff
@@ -202,6 +202,8 @@ bitflags::bitflags! {
         const STORAGE_BUFFER_BINDING_ARRAY_NON_UNIFORM_INDEXING = 1 << 35;
         /// Support for cooperative matrix types and operations
         const COOPERATIVE_MATRIX = 1 << 36;
+        /// Support for per-vertex fragment input.
+        const SHADER_PER_VERTEX = 1 << 37;
```
Comment (inner-daemons) [inline naga/src/valid/mod.rs]: Maybe this should be called `PER_VERTEX` or `PER_VERTEX_FRAGMENT_INPUT` or something. These are already shader capabilities so `SHADER_` is unnecessary

File: wgpu-types/src/features.rs [hunk 2]
```diff
@@ -1274,6 +1274,14 @@ bitflags_array! {
         ///
         /// This is a native only feature.
         const EXPERIMENTAL_COOPERATIVE_MATRIX = 1 << 57;
+
+        /// Enables shader barycentric coordinates.
```
Comment (inner-daemons) [inline wgpu-types/src/features.rs]: Forgot to document this different than the barycentric feature

File: CHANGELOG.md [hunk 3]
```diff
@@ -44,6 +44,7 @@ Bottom level categories:
 ### New Features
 
 - Added support for cooperative load/store operations in shaders. Currently only WGSL on the input and SPIR-V, METAL, and WGSL on the output are supported. By @kvark in [#8251](https://github.com/gfx-rs/wgpu/issues/8251).
+- Added support for per-vertex attributes in fragment shaders. Currently only WGSL on the input and SPIR-V, and WGSL on the output are supported. By @atlv in [#8821](https://github.com/gfx-rs/wgpu/issues/8821).
```
Comment (inner-daemons) [inline CHANGELOG.md]: Probably could be worded better, particularly "WGSL on the input and SPIR-V, and WGSL on the output"

File: wgpu-hal/src/vulkan/adapter.rs [hunk 4]
```diff
@@ -2341,7 +2345,9 @@ impl super::Adapter {
                 capabilities.push(spv::Capability::ClipDistance);
             }
 
-            if features.intersects(wgt::Features::SHADER_BARYCENTRICS) {
+            if features
+                .intersects(wgt::Features::SHADER_BARYCENTRICS | wgt::Features::SHADER_PER_VERTEX)
```
Comment (inner-daemons) [inline wgpu-hal/src/vulkan/adapter.rs:2351]: Maybe add a comment for why either of these should pass the capability

File: wgpu-hal/src/vulkan/adapter.rs [hunk 5]
```diff
@@ -1293,7 +1295,9 @@ impl PhysicalDeviceProperties {
         }
 
         // Require `VK_KHR_fragment_shader_barycentric` if the associated feature was requested
-        if requested_features.intersects(wgt::Features::SHADER_BARYCENTRICS) {
+        if requested_features
+            .intersects(wgt::Features::SHADER_BARYCENTRICS | wgt::Features::SHADER_PER_VERTEX)
+        {
```
Comment (inner-daemons) [inline wgpu-hal/src/vulkan/adapter.rs:1301]: Same as the comment below

File: tests/tests/wgpu-gpu/per_vertex/mod.rs
Comment (inner-daemons) [inline tests/tests/wgpu-gpu/per_vertex/mod.rs:1]: It might not be necessary for this PR, but should we add a test using triangle strip topology to better catch potential edge cases with provoking vertices?

File: naga/tests/naga/validation.rs
Comment (inner-daemons) [inline naga/tests/naga/validation.rs:1]: Appreciate this testing!

File: naga/src/valid/interface.rs [hunk 6]
```diff
@@ -443,8 +445,18 @@ impl VaryingContext<'_> {
                         Capabilities::MESH_SHADER,
                     ));
                 }
+                if interpolation == Some(crate::Interpolation::PerVertex) {
+                    if !self.capabilities.contains(Capabilities::SHADER_PER_VERTEX)
+                        || self.stage != crate::ShaderStage::Fragment
+                    {
+                        return Err(VaryingError::UnsupportedCapability(
+                            Capabilities::SHADER_PER_VERTEX,
+                        ));
```
Comment (inner-daemons) [inline naga/src/valid/interface.rs:464]: We should have a different error for usage outside of fragment shaders. This doesn't really have to do with capabilities

File: naga/src/valid/interface.rs [hunk 7]
```diff
@@ -443,8 +445,18 @@ impl VaryingContext<'_> {
...
+                    }
+                }
                 // Only IO-shareable types may be stored in locations.
-                if !self.type_info[ty.index()]
+                // Per Vertex case is checked later.
+                else if !self.type_info[ty.index()]
```
Comment (inner-daemons) [inline naga/src/valid/interface.rs]: Doesn't look like we actually check the type inside the array

File: naga/src/front/spv/mod.rs
Comment (inner-daemons) [inline naga/src/front/spv/mod.rs:1]: Can we get testing for spv-in if we are gonna add parsing support?

**Review by inner-daemons — CHANGES_REQUESTED** [review]
Broadly looks good. Some comments for now, plus I want to think through the validation later to make sure you didn't miss anything.

File: naga/src/valid/interface.rs (hunk 7 above)
Comment (atlv24) [reply naga/src/valid/interface.rs]: It is checked here:
```rs
if interpolation == Some(crate::Interpolation::PerVertex) {
    let three = crate::ArraySize::Constant(core::num::NonZeroU32::new(3).unwrap());
    match ty_inner {
        &Ti::Array { base, size, .. } if size == three => {
            if self.types[base].inner.scalar_kind().is_none() {
                return Err(VaryingError::InvalidType(base));
            }
        }
        _ => return Err(VaryingError::PerVertexNotArrayOfThree),
    }
}
```

File: naga/src/front/spv/mod.rs
Comment (atlv24) [reply naga/src/front/spv/mod.rs:1]: I am not supporting spv-in. wgsl-in is tested via naga tests i believe

Comment (inner-daemons) [reply naga/src/front/spv/mod.rs:1]: Why have you modified the spirv frontend then to support `PerVertexKHR`? And if there's nothing else to do, doesn't that count as supporting spv-in?

File: naga/src/valid/interface.rs (hunk 7 above)
Comment (inner-daemons) [reply naga/src/valid/interface.rs]: I think for organizational purposes we should keep the checks nearby since they should be identical. Also, why did you use the same method of checking for `IO_SHARABLE` flag?

File: wgpu-hal/src/vulkan/adapter.rs [hunk 8]
```diff
@@ -1292,8 +1294,11 @@ impl PhysicalDeviceProperties {
             extensions.push(ext::mesh_shader::NAME);
         }
 
-        // Require `VK_KHR_fragment_shader_barycentric` if the associated feature was requested
-        if requested_features.intersects(wgt::Features::SHADER_BARYCENTRICS) {
+        // Require `VK_KHR_fragment_shader_barycentric` if an associated feature was requested
+        // Vulkan bundles both barycentrics nad per-vertex attributes under the same feature.
```
[the remark this answers is not in the document]
Comment (inner-daemons) [reply wgpu-hal/src/vulkan/adapter.rs]: Amazing use of 50 tons of coal

File: naga/src/front/spv/mod.rs
Comment (atlv24) [reply naga/src/front/spv/mod.rs:1]: OH i see, i misunderstood. What kind of test would this entail?

Comment (inner-daemons) [reply naga/src/front/spv/mod.rs:1]: Get some spir-v that uses per-vertex data and put it in a .spvasm file in naga/tests/in/spv. Can be generated from glslang, naga, etc, doesn't really matter. Just a simple snapshot

File: naga/src/valid/interface.rs [hunk 9]
```diff
@@ -58,6 +58,8 @@ pub enum VaryingError {
     NotIOShareableType(Handle<crate::Type>),
     #[error("Interpolation is not valid")]
     InvalidInterpolation,
+    #[error("Interpolation {0:?} is only valid for stage {1:?}")]
+    InvalidInterpolationInStage(crate::Interpolation, crate::ShaderStage),
```
Comment (inner-daemons) [inline naga/src/valid/interface.rs:62]: Thanks!

File: naga/src/valid/interface.rs [hunk 10]
```diff
@@ -550,19 +568,32 @@ impl VaryingContext<'_> {
...
-                    Some(crate::ScalarKind::Float) => {
-                        if needs_interpolation && interpolation.is_none() {
-                            return Err(VaryingError::MissingInterpolation);
+                if interpolation == Some(crate::Interpolation::PerVertex) {
+                    let three = crate::ArraySize::Constant(core::num::NonZeroU32::new(3).unwrap());
+                    match ty_inner {
+                        &Ti::Array { base, size, .. } if size == three => {
+                            if self.types[base].inner.scalar_kind().is_none() {
+                                return Err(VaryingError::InvalidType(base));
+                            }
```
Comment (inner-daemons) [inline naga/src/valid/interface.rs]: Previous comment got resolved so I'm commenting here:

I think for organizational purposes we should keep the checks for IO_SHARABLE types nearby since they should be identical. Also, why didn't you use the same method of checking for IO_SHARABLE flag, instead using this scalar_kind stuff?

File: tests/tests/wgpu-gpu/per_vertex/mod.rs
Comment (inner-daemons) [inline tests/tests/wgpu-gpu/per_vertex/mod.rs:1]: Could we make the texture a little bit bigger just to be perfectly clear with whats going on? Maybe just like 16x16 or something

**Review by inner-daemons — CHANGES_REQUESTED** [review]
Pretty much good I think! Thanks for working on this. A few more questions, nothing major

File: tests/tests/wgpu-gpu/per_vertex/mod.rs
Comment (atlv24) [reply tests/tests/wgpu-gpu/per_vertex/mod.rs:1]: I believe the point of the other tests using a small texture is to be able to reason about it pixel by pixel mentally. With that many pixels, you'd be having to reason about 256 fragment shader invocations mentally which is a bit much. It would also make the assert huge and unreadable with 1024 byte literals

File: naga/src/valid/interface.rs (hunk 10 above)
Comment (atlv24) [reply naga/src/valid/interface.rs]: When i wrote this PR the IO_SHARABLE type flag did not exist. Either way i don't think its the right check? Looking at IO_SHARABLE's doc comment,
```
/// This covers anything that can be in [`Location`] binding:
/// non-bool scalars and vectors, matrices, and structs and
/// arrays containing only interface types.
```
I don't think structs and arrays containing interface types can be in a per-vertex array in a meaningful sense, but maybe im wrong. I took the check for io-sharable types at the time and used only the scalar/vector/matrix path.

File: naga/src/front/spv/mod.rs
Comment (atlv24) [reply naga/src/front/spv/mod.rs:1]: added

File: naga/src/valid/interface.rs [hunk 11]
```diff
@@ -443,6 +447,31 @@ impl VaryingContext<'_> {
...
+                            crate::ShaderStage::Fragment,
+                        ));
+                    }
+                    if !self.capabilities.contains(Capabilities::PER_VERTEX) {
+                        return Err(VaryingError::UnsupportedCapability(
+                            Capabilities::PER_VERTEX,
+                        ));
+                    }
+                }
+                let (ty, ty_inner) = if interpolation == Some(crate::Interpolation::PerVertex) {
```
Comment (cwfitzgerald) [inline naga/src/valid/interface.rs:469]: Could you add a comment here that we're changing the type we're dealing with to be the _inner_ type if there's a per-vertex, otherwise leaving it be - so all validation is done on the inner type, now that we've asserted it's array<T, 3>

File: naga/tests/out/ir/spv-per-vertex.compact.ron
Comment (cwfitzgerald) [inline naga/tests/out/ir/spv-per-vertex.compact.ron:1]: As we have output to real backends, lets remove the ron outputs to make the snapshots a bit less noisy.

File: tests/tests/wgpu-gpu/per_vertex/mod.rs
Comment (cwfitzgerald) [reply tests/tests/wgpu-gpu/per_vertex/mod.rs:1]: I agree that this size is perfect, the diagram really helps!

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Code looks good!

Some smaller nits, plus could we add a test in wgsl_error to make sure that any per_vertex stuff gets properly validated out if the feature isn't there (very important for firefox). This would also be a good place to test the other validation, but this is less important.

**Review by cwfitzgerald — APPROVED** [review]
I like to merge it merge it
